### PR TITLE
Merge aspnet-telemetrycorrelation branch back to main

### PR DESCRIPTION
**Pull request by reyang**
Part of #2222.

Note: we need to merge this PR by keeping all the commits rather than squashing them (so we know what's been copied over directly vs. changed by the OpenTelemetry .NET community).
Note: I didn't include the changelog yet, we might want to add a single entry with all the related PRs, similar like what we've done in the spec repo https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md#metrics-2.

@CodeBlanch I think we have two options, want to get your opinion:
1. merge this to main right now, and you will continue the work in main branch (with the benefit of CI, less chance of hitting merge conflict when we touch the ASP.NET instrumentation lib, etc.)
2. keep this in the `aspnet-telemetrycorrelation` branch, which means you will continue the refactor (including renaming the package/namespace and the integration with ASP.NET instrumentation), and once you finished, merge this PR (which will be bigger at that time) to main.

### CODE REVIEW

**Review by cijothomas — APPROVED** [review]

**Comment by cijothomas** [comment]
marking as "donotmerge" until we temporarily enable merge-with-keep-commits.

**Comment by reyang** [comment]
> marking as "donotmerge" until we temporarily enable merge-with-keep-commits.

And we want to wait to get approval (or rejection) from @CodeBlanch as he will be taking it further after I finished the grunt work.

**Comment by cijothomas** [comment]
> I think we're good to merge this now. @cijothomas What do we do with API Compatability check on new projects?

Similar to the below:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Prometheus/OpenTelemetry.Exporter.Prometheus.csproj#L14-L18

**Pull request merged**
